### PR TITLE
cwrappers for ntheory

### PR DESCRIPTION
**Pull request by kunal-iitkgp**
wrapped the following functions from ntheory
gcd
lcm
nextprime
mod
quotient
fibonacci
lucas
binomial

also added the test cases in `test_cwrapper.c`
 @certik @isuruf @abinashmeher999  pls review

### CODE REVIEW

File: symengine/cwrapper.h [hunk 1]
```diff
@@ -256,9 +256,26 @@ void basic_subs2(basic s, const basic e, const basic a, const basic b);
...
+//! Least Common Multiple
+void ntheory_lcm(basic s, const basic a, const basic b);
+//! \return next prime after `a`
+void ntheory_nextprime(basic s, const basic a);
+//! modulo round toward zero
+void ntheory_mod(basic s, const basic n, const basic d);
+//! \return quotient round toward zero when `n` is divided by `d`
+void ntheory_quotient(basic s, const basic n, const basic d);
+//! Fibonacci number
+void ntheory_fibonacci(basic s, unsigned long a);
```
Comment (abinashmeher999) [inline symengine/cwrapper.h]: Add a clarification in the documentation where the sequence starts from, 0 or 1, and also if we are counting with 0 index as first element or 1 index as the first.

File: symengine/tests/cwrapper/test_cwrapper.c [hunk 2]
```diff
@@ -458,6 +458,69 @@ void test_ascii_art() {
...
+    basic_free_stack(x5);
+    basic_free_stack(x6);
+    basic_free_stack(x7);
+    basic_free_stack(x8);
+    basic_free_stack(i1);
+    basic_free_stack(i2);
+    basic_free_stack(i4);
+    basic_free_stack(i5);
+
+
```
Comment (abinashmeher999) [inline symengine/tests/cwrapper/test_cwrapper.c]: It would be good to get rid of these blank newlines.

File: symengine/cwrapper.cpp [hunk 3]
```diff
@@ -489,6 +489,47 @@ char* ascii_art_str()
     return cc;
 }
 
+
+void ntheory_gcd(basic s, const basic a, const basic b)
+{
+    s->m = SymEngine::gcd(static_cast<const Integer &>(*(a->m)), static_cast<const Integer &>(*(b->m)));
```
Comment (abinashmeher999) [inline symengine/cwrapper.cpp]: I am not sure if first dereferencing and then casting is the right way to do it when using `RCP`. You might want to use [`rcp_static_cast`](https://github.com/symengine/symengine/blob/master/symengine/symengine_rcp.h#L146) or [`rcp_dynamic_cast`](https://github.com/symengine/symengine/blob/master/symengine/symengine_rcp.h#L154) to first cast the `RCP` and then dereference it.

Comment (isuruf) [inline symengine/cwrapper.cpp]: Nope. This is correct

File: symengine/cwrapper.cpp [hunk 4]
```diff
@@ -489,6 +489,47 @@ char* ascii_art_str()
...
+}
+
+void ntheory_lucas(basic s, unsigned long a)
+{
+    s->m = SymEngine::lucas(a);
+}
+
+void ntheory_binomial(basic s, const basic a, unsigned long b)
+{
+    s->m = SymEngine::binomial(static_cast<const Integer &>(*(a->m)), b);
```
Comment (isuruf) [inline symengine/cwrapper.cpp]: You should never cast to a type you are not sure of. At least a debug time check is needed before casting to a type. See, `integer_get_mpz` for an example

File: symengine/tests/cwrapper/test_cwrapper.c [hunk 5]
```diff
@@ -458,6 +458,69 @@ void test_ascii_art() {
     basic_str_free(s);
 }
 
+void test_ntheory() {
+    basic x1, x2, x3, x4, x5, x6, x7, x8,  i1, i2, i4, i5;
```
Comment (isuruf) [inline symengine/tests/cwrapper/test_cwrapper.c]: You don't need all the variables `x1, x2, x3, x4, x5, x6, x7, x8`. Just use one `x`

**Comment by kunal-iitkgp** [comment]
ping @isuruf  @abinashmeher999

File: symengine/tests/cwrapper/test_cwrapper.c [hunk 6]
```diff
@@ -458,6 +458,54 @@ void test_ascii_art() {
...
+    basic_free_stack(x);
+    basic_free_stack(i1);
+    basic_free_stack(i2);
+    basic_free_stack(i4);
+    basic_free_stack(i5);
+
+}
+
+
+
```
Comment (isuruf) [inline symengine/tests/cwrapper/test_cwrapper.c]: Three empty lines are unnecessary

File: symengine/cwrapper.cpp (hunk 3 above)
Comment (abinashmeher999) [inline symengine/cwrapper.cpp]: @isuruf Should this be refactored? https://github.com/symengine/symengine/pull/852/files#diff-72450588351eec10a19415c0a981ad41R177 Maybe in another PR.

File: symengine/tests/cwrapper/test_cwrapper.c (hunk 6 above)
Comment (kunal-iitkgp) [inline symengine/tests/cwrapper/test_cwrapper.c]: thanks. will do away with them

File: symengine/tests/cwrapper/test_cwrapper.c [hunk 7]
```diff
@@ -458,6 +458,54 @@ void test_ascii_art() {
     basic_str_free(s);
 }
 
+void test_ntheory() {
+    basic x,  i1, i2, i4, i5;
```
Comment (abinashmeher999) [inline symengine/tests/cwrapper/test_cwrapper.c]: 2 spaces here.

File: symengine/cwrapper.h [hunk 8]
```diff
@@ -256,9 +256,26 @@ void basic_subs2(basic s, const basic e, const basic a, const basic b);
...
+//! nth Fibonacci number //  fibonacci(0) = 0 and fibonacci(1) = 1
+void ntheory_fibonacci(basic s, unsigned long a);
+//! Lucas number
+void ntheory_lucas(basic s, unsigned long a);
+//! Binomial Coefficient
+void ntheory_binomial(basic s, const basic a, unsigned long b);
 //! Print stacktrace on segfault
 void symengine_print_stack_on_segfault();
 
+
```
Comment (isuruf) [inline symengine/cwrapper.h]: This is not needed

File: symengine/cwrapper.h [hunk 9]
```diff
@@ -256,9 +256,26 @@ void basic_subs2(basic s, const basic e, const basic a, const basic b);
...
+//! modulo round toward zero
+void ntheory_mod(basic s, const basic n, const basic d);
+//! \return quotient round toward zero when `n` is divided by `d`
+void ntheory_quotient(basic s, const basic n, const basic d);
+//! nth Fibonacci number //  fibonacci(0) = 0 and fibonacci(1) = 1
+void ntheory_fibonacci(basic s, unsigned long a);
+//! Lucas number
+void ntheory_lucas(basic s, unsigned long a);
+//! Binomial Coefficient
+void ntheory_binomial(basic s, const basic a, unsigned long b);
```
Comment (isuruf) [inline symengine/cwrapper.h]: Add space after this line

**Comment by kunal-iitkgp** [comment]
any more corrections  required ?? @isuruf @abinashmeher999

**Comment by abinashmeher999** [comment]
Looks good to me. It would be good if you could squash the cleanup commits like `fixed typos`, `removed empty lines` etc with other ones.

**Comment by kunal-iitkgp** [comment]
squashed `removed empty spaces`, `removed empty lines` and `managed newlines ` into `added integer check`

**Comment by abinashmeher999** [comment]
+1 to merge. @isuruf ?

**Comment by isuruf** [comment]
+1 from me as well

**Pull request merged**